pkg/internal/source/kind.go

```diff
@@ -24,72 +27,142 @@ type Kind struct {
+		return fmt.Errorf("cannot start an already started Kind source")
+	}
+
+	ctx, ks.cancel = context.WithCancel(ctx)
+
+	ks.started = make(chan error, 1)
 	go func() {
-		var (
-			i       cache.Informer
-			lastErr error
+		defer close(ks.started)
+
+		ks.started <- ks.registerEventHandler(
+			ctx,
+			NewEventHandler(ctx, queue, handler, prct).HandlerFuncs(),
 		)
+	}()
 
-		// Tries to get an informer until it returns true,
-		// an error or the specified context is cancelled or expired.
-		if err := wait.PollImmediateUntilWithContext(ctx, 10*time.Second, func(ctx context.Context) (bool, error) {
-			// Lookup the Informer from the Cache and add an EventHandler which populates the Queue
-			i, lastErr = ks.Cache.GetInformer(ctx, ks.Type)
-			if lastErr != nil {
-				kindMatchErr := &meta.NoKindMatchError{}
-				switch {
-				case errors.As(lastErr, &kindMatchErr):
-					log.Error(lastErr, "if kind is a CRD, it should be installed before calling Start",
-						"kind", kindMatchErr.GroupKind)
-				case runtime.IsNotRegisteredError(lastErr):
-					log.Error(lastErr, "kind must be registered to the Scheme")
-				default:
-					log.Error(lastErr, "failed to get informer from cache")
-				}
-				return false, nil // Retry.
-			}
-			return true, nil
-		}); err != nil {
-			if lastErr != nil {
-				ks.started <- fmt.Errorf("failed to get informer from cache: %w", lastErr)
-				return
+	return nil
+}
+
+func getInformer(ctx context.Context, informerCache cache.Cache, resourceType client.Object) (cache.Informer, error) {
+	var (
+		i       cache.Informer
+		lastErr error
+	)
+
+	// Tries to get an informer until it returns true,
+	// an error or the specified context is cancelled or expired.
+	if err := wait.PollImmediateUntilWithContext(ctx, 10*time.Second, func(pollCtx context.Context) (bool, error) {
+		// Lookup the Informer from the Cache and add an EventHandler which populates the Queue
+		i, lastErr = informerCache.GetInformer(pollCtx, resourceType)
+		if lastErr != nil {
+			kindMatchErr := &meta.NoKindMatchError{}
+			switch {
+			case errors.As(lastErr, &kindMatchErr):
+				// We got a NoKindMatchError, which means the kind is a CRD and it's not installed yet.
+				// We should retry until it's installed.
+				log.Error(lastErr, "if kind is a CRD, it should be installed before calling Start", "kind", kindMatchErr.GroupKind)
+			case runtime.IsNotRegisteredError(lastErr):
+				// We got a IsNotRegisteredError, which means the kind is not registered to the Scheme.
+				// This is a programming error, so we should stop retrying and return the error.
+				return true, fmt.Errorf("kind must be registered to the Scheme: %w", lastErr)
+			default:
+				// We got an error that is not a NoKindMatchError or IsNotRegisteredError, so we should
+				// stop retrying and return the error.
+				return true, fmt.Errorf("failed to get informer from cache: %w", lastErr)
 			}
-			ks.started <- err
-			return
+			return false, nil // Retry.
```

Comment:
nit: I would suggest to move this line after l.103. Just slightly more readable vs the fallthrough behavior